c10/util/complex.h

```diff
@@ -0,0 +1,485 @@
+    storage[0] += rhs.real();
+    storage[1] += rhs.imag();
+    return static_cast<complex<T> &>(*this);
+  }
+
+  template<typename U>
+  constexpr complex<T> &operator -=(const complex<U> &rhs) {
+    storage[0] -= rhs.real();
+    storage[1] -= rhs.imag();
+    return static_cast<complex<T> &>(*this);
+  }
+
+  template<typename U>
+  constexpr complex<T> &operator *=(const complex<U> &rhs) {
+    // (a + bi) * (c + di) = (a*c - b*d) + (a * d + b * c) i
+    T a = storage[0];
+    T b = storage[1];
+    U c = rhs.real();
+    U d = rhs.imag();
+    storage[0] = a * c - b * d;
+    storage[1] = a * d + b * c;
+    return static_cast<complex<T> &>(*this);
+  }
+
+  template<typename U>
+  constexpr complex<T> &operator /=(const complex<U> &rhs) {
+    // (a + bi) / (c + di) = (ac + bd)/(c^2 + d^2) + (bc - ad)/(c^2 + d^2) i
+    T a = storage[0];
+    T b = storage[1];
+    U c = rhs.real();
+    U d = rhs.imag();
+    auto denominator = c * c + d * d;
+    storage[0] = (a * c + b * d) / denominator;
+    storage[1] = (b * c - a * d) / denominator;
+    return static_cast<complex<T> &>(*this);
+  }
+
+  template<typename U>
+  constexpr complex<T> &operator =(const std::complex<U> &rhs) {
+    storage[0] = rhs.real();
+    storage[1] = rhs.imag();
+    return static_cast<complex<T> &>(*this);
+  }
+
+#if defined(__CUDACC__) || defined(__HIPCC__)
+  template<typename U>
+  C10_HOST_DEVICE complex<T> &operator =(const thrust::complex<U> &rhs) {
+    storage[0] = rhs.real();
+    storage[1] = rhs.imag();
+    return static_cast<complex<T> &>(*this);
+  }
+#endif
+
+  template<typename U>
+  explicit constexpr operator std::complex<U>() const {
+    return std::complex<U>(std::complex<T>(real(), imag()));
+  }
+
+#if defined(__CUDACC__) || defined(__HIPCC__)
+  template<typename U>
+  explicit operator thrust::complex<U>() const {
+    return thrust::complex<U>(thrust::complex<T>(real(), imag()));
+  }
+#endif
+
+  constexpr T real() const {
+    return storage[0];
+  }
+  constexpr void real(T value) {
+    storage[0] = value;
+  }
+  constexpr T imag() const {
+    return storage[1];
+  }
+  constexpr void imag(T value) {
+    storage[1] = value;
+  }
+};
+
+
+template<>
+struct alignas(2*sizeof(float)) complex<float>: public complex_common<float> {
+  using complex_common<float>::complex_common;
+  constexpr complex(): complex_common() {}; // needed by CUDA 9.x
+  explicit constexpr complex(const complex<double> &other);
+  using complex_common<float>::operator=;
+};
+
+template<>
+struct alignas(2*sizeof(double)) complex<double>: public complex_common<double> {
+  using complex_common<double>::complex_common;
+  constexpr complex(): complex_common() {}; // needed by CUDA 9.x
+  constexpr complex(const complex<float> &other);
+  using complex_common<double>::operator=;
+};
+
+constexpr complex<float>::complex(const complex<double> &other): complex_common(other.real(), other.imag()) {}
```

Comment:
I think these methods should be explicit because this:
```
std::complex<double> a(0,1);
std::complex<float> b = a;
``` 
should fail

Reply:
`explicit` can only be used inside the class definition. This function is *declared* as explicit inside the class, and *defined* outside. When defining, we can not add `explicit`